src/hammer-tech/stackup.py

```diff
@@ -0,0 +1,198 @@
+        spacing = ws[0].min_spacing
+        for first, second in zip(ws[:-1], ws[1:]):
+            if s2w >= (2*first.min_spacing + second.width_at_least):
+                spacing = second.min_spacing
+        width = s2w - spacing*2
+        # TODO this is assuming that a manufacturing grid is 0.001
+        assert (int(self.min_width * 1000) % 2 == 0), "Assuming all min widths are even here, if not fix me"
+        start = self.min_width / 2 + spacing
+        return (width, spacing, start)
+
+    # This method will return the maximum width a wire can be to consume a given number of routing tracks
+    # This assumes both neighbors are wires of the same width
+    # i.e. W W W
+    # T = thin / min-width
+    # W = wide
+    # Returns width, spacing, and start
+    def get_width_spacing_start_www(self, tracks: int, force_even=False) -> Tuple[float, float, float]:
+        raise NotImplementedError("Not yet implemented")
+
+    # This method will return the maximum width a wire can be to consume a given number of routing tracks
+    # This assumes one neighbor is a min width wire, and the other is the same size as this (mirrored)
+    # i.e. T W W T
+    # T = thin / min-width
+    # W = wide
+    # Returns width, spacing, and offset
+    # The offset is the offset of the wire centerline to the track (odd number of tracks) or half-track (even number of tracks)
+    # Positive numbers towards min-width wire
+    def get_width_spacing_start_twwt(self, tracks: int, force_even=False) -> Tuple[float, float, float]:
+        ws = self.power_strap_widths_and_spacings
+        s3w2 = (2 * tracks + 1) * self.pitch - self.min_width
+        spacing = ws[0].min_spacing
+        for first, second in zip(ws[:-1], ws[1:]):
+            if s3w2 >= (3*first.min_spacing + 2*second.width_at_least):
+                spacing = second.min_spacing
+        width = (s3w2 - spacing*3)/2
+        start = self.min_width / 2 + spacing
+        # TODO this is assuming that a manufacturing grid is 0.001
+        # Note the rounding here seems to get around some floating-point roundoff issues in prior calculations
+        if force_even and ((int(round(width * 1000)) % 2) != 0):
+                width = width - 0.001
+                start = start + 0.001
+        return (width, spacing, start)
+
+    # TODO implement M W X* W M style wires, where X is slightly narrower than W and centered on-grid
+
+# For now a stackup is just a list of metals with a meaningful keyword name
+# TODO add vias, etc. when we need them
+class Stackup(NamedTuple('Stackup', [
+    ('name', str),
+    ('metals', List[Metal])
+])):
+    __slots__ = ()
+
+    @staticmethod
+    def from_setting(d: dict) -> "Stackup":
+        return Stackup(
+            name=str(d["name"]),
+            metals=list(map(lambda x: Metal.from_setting(x), list(d["metals"])))
+        )
+
+
+    def get_metal(self, name: str) -> "Metal":
+        # TODO don't brute force this
```

Comment:
Not sure what is meant by this; is this about writing it in a more functional way?

Reply:
Yes, I deleted the comment. I think it's fine as-is.